keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/LogProbGradients.java

```diff
@@ -0,0 +1,53 @@
+package io.improbable.keanu.vertices.dbl.nonprobabilistic.diff;
+
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.vertices.VertexId;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * This class stores the gradients of a log probability. It serves
+ * as way to sum multiple gradients from the same vertices.
+ */
+public class LogProbGradients {
+
+    private final Map<VertexId, DoubleTensor> partials;
+
+    public LogProbGradients() {
+        this.partials = new HashMap<>();
+    }
+
+    public LogProbGradients add(LogProbGradients addition) {
+        return add(addition.partials);
+    }
+
+    public LogProbGradients add(Map<VertexId, DoubleTensor> addition) {
+
+        for (Map.Entry<VertexId, DoubleTensor> entry : addition.entrySet()) {
+
+            VertexId id = entry.getKey();
+            DoubleTensor existingPartialDerivative = partials.get(id);
+
+            if (existingPartialDerivative == null) {
+                partials.put(id, entry.getValue().duplicate());
+            } else {
+                existingPartialDerivative.plusInPlace(entry.getValue());
```

Comment:
Don't we have a potential bug hiding here now?  plusInPlace sometimes doesn't actually do things in place...

Reply:
At the moment it's impossible for the existing partial to have a different shape to the next partial. All partials at this point are the correct deterministic shape [of, wrt]. That being said, the inPlace contract doesn't guarantee that the object will always be the same. I've gone with the double put to make sure this is never an issue.

Reply:
I think better to be safe than sorry definitely!